pkg/ovs/openflow/ofctrl_action.go

```diff
@@ -516,87 +510,87 @@ func (a *ofLearnAction) MatchTransportDst(protocol Protocol) LearnAction {
-// must match the sctp_dst of the packet currently being processed.
-func (a *ofLearnAction) MatchLearnedSCTPv6DstPort() LearnAction {
-	return a.MatchTransportDst(ProtocolSCTPv6)
-}
-
-// MatchLearnedSrcIP makes the learned flow to match the nw_src of current IP packet.
-func (a *ofLearnAction) MatchLearnedSrcIP() LearnAction {
-	a.nxLearn.AddMatch(&ofctrl.LearnField{Name: "NXM_OF_IP_SRC"}, 4*8, &ofctrl.LearnField{Name: "NXM_OF_IP_SRC"}, nil)
+// MatchLearnedDstPort specifies that the transport layer destination field
+// {tcp|udp|sctp}_dst in the learned flow must match the same field of the packet
+// currently being processed. It only accepts ProtocolTCP, ProtocolUDP, or
+// ProtocolSCTP, otherwise this does nothing.
+func (a *ofLearnAction) MatchLearnedDstPort(protocol Protocol) LearnAction {
+	// OXM_OF fields support TCP, UDP and SCTP, but NXM_OF fields only support TCP and UDP. So here using "OXM_OF_" to
+	// generate the field name.
+	var regName string
+	switch protocol {
+	case ProtocolTCP, ProtocolTCPv6:
+		regName = OxmFieldTCPDst
+	case ProtocolUDP, ProtocolUDPv6:
+		regName = OxmFieldUDPDst
+	case ProtocolSCTP, ProtocolSCTPv6:
+		regName = OxmFieldSCTPDst
+	default:
+		// Return directly if the protocol is not acceptable.
+		return a
+	}
+	a.nxLearn.AddMatch(&ofctrl.LearnField{Name: regName}, 2*8, &ofctrl.LearnField{Name: regName}, nil)
 	return a
 }
 
-// MatchLearnedDstIP makes the learned flow to match the nw_dst of current IP packet.
-func (a *ofLearnAction) MatchLearnedDstIP() LearnAction {
-	a.nxLearn.AddMatch(&ofctrl.LearnField{Name: "NXM_OF_IP_DST"}, 4*8, &ofctrl.LearnField{Name: "NXM_OF_IP_DST"}, nil)
+// MatchLearnedSrcPort specifies that the transport layer source field
+// {tcp|udp|sctp}_src in the learned flow must match the same field of the packet
+// currently being processed. It only accepts ProtocolTCP, ProtocolUDP, or
+// ProtocolSCTP, otherwise this does nothing.
+func (a *ofLearnAction) MatchLearnedSrcPort(protocol Protocol) LearnAction {
+	// OXM_OF fields support TCP, UDP and SCTP, but NXM_OF fields only support TCP and UDP. So here using "OXM_OF_" to
+	// generate the field name.
+	var regName string
+	switch protocol {
+	case ProtocolTCP, ProtocolTCPv6:
+		regName = OxmFieldTCPSrc
+	case ProtocolUDP, ProtocolUDPv6:
+		regName = OxmFieldUDPSrc
+	case ProtocolSCTP, ProtocolSCTPv6:
+		regName = OxmFieldSCTPSrc
+	default:
+		// Return directly if the protocol is not acceptable.
+		return a
+	}
+	a.nxLearn.AddMatch(&ofctrl.LearnField{Name: regName}, 2*8, &ofctrl.LearnField{Name: regName}, nil)
 	return a
 }
 
-// MatchLearnedSrcIPv6 makes the learned flow to match the ipv6_src of current IPv6 packet.
-func (a *ofLearnAction) MatchLearnedSrcIPv6() LearnAction {
-	a.nxLearn.AddMatch(&ofctrl.LearnField{Name: "NXM_NX_IPV6_SRC"}, 16*8, &ofctrl.LearnField{Name: "NXM_NX_IPV6_SRC"}, nil)
+// MatchLearnedSrcIP makes the learned flow to match the nw_src of current IP packet.
```

Comment:
"flow match nw_src of the current IP packet"

Reply:
Done